Fix for regression in issue #6472

JFCote opened the pull request:
Fix regression explained here: #6472 

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the correct branch: master for non-breaking changes and `3.0.0` branch for breaking (non-backward compatible) changes.

### Description of the PR

I only run Java Play Framework .sh files. @daveish , can you try running your samples and we will see if everything compiles properly.

Thanks

wing328 commented:
@JFCote I've run ./bin/java-petstore-all.sh to update all Java API clients. CircleCI tests look good: https://circleci.com/gh/swagger-api/swagger-codegen/1658

@daveish may I know which generator you'd issue with before? can you also run some tests to ensure it's not breaking your use case?
```
git checkout -b StingrayDigital-redo-java8-diamond-fix master
git pull https://github.com/StingrayDigital/swagger-codegen.git redo-java8-diamond-fix
mvn clean package
# generate the client/server
```

daveish commented:
I ran `./bin/jaxrs-resteasy-eap-java8-petstore-server.sh`

I get this for CircleCI tests:

```[ERROR] COMPILATION ERROR : 
[ERROR] /home/david/git/swagger-codegen/samples/server/petstore/jaxrs-resteasy/eap-java8/src/gen/java/io/swagger/model/Pet.java:[22,50] diamond operator is not supported in -source 1.5
  (use -source 7 or higher to enable diamond operator)
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.2:compile (default-compile) on project swagger-jaxrs-resteasy-eap-java8-server: Compilation failure
[ERROR] /home/david/git/swagger-codegen/samples/server/petstore/jaxrs-resteasy/eap-java8/src/gen/java/io/swagger/model/Pet.java:[22,50] diamond operator is not supported in -source 1.5
[ERROR] (use -source 7 or higher to enable diamond operator)```

JFCote commented:
It seems you try to compile using java 1.5 but you generates your server for java 8.  Can you validate which version of java you have on your computer by running `java -version`

daveish commented:
Java8 is the only version i ever installed, but here is my output:

```java version "1.8.0_144"
Java(TM) SE Runtime Environment (build 1.8.0_144-b01)
Java HotSpot(TM) 64-Bit Server VM (build 25.144-b01, mixed mode)```

JFCote commented:
@daveish : So there is a language level flag somewhere. I don't know if it's in the generated code or somewhere else. It seems to be set to 1.5 which prevent the usage of java 8 functionalities even if you have the right jdk... 

Maybe @wing328 can help here?

daveish commented:
I can't say anything there. I did just execute the sample generator, I did not rewrite any logic for this sample. This sample generator was just not executed in previous commits, I guess.

JFCote commented:
@daveish 

I just run the ./bin/jaxrs-resteasy-eap-java8-petstore-server.sh on the StingrayDigital/swagger-codegen/redo-java8-diamond-fix and have no error. Here is the output (there is no errors).

```
$ ./bin/jaxrs-resteasy-eap-java8-petstore-server.sh
Removing files and folders under samples/server/petstore/jaxrs-resteasy/eap-java8/src/main
[main] INFO io.swagger.parser.Swagger20Parser - reading from modules/swagger-codegen/src/test/resources/2_0/petstore.yaml
[main] INFO io.swagger.codegen.ignore.CodegenIgnoreProcessor - No .swagger-codegen-ignore file found.
[main] INFO io.swagger.codegen.DefaultCodegen - Skipped overwriting README.md as the file already exists in C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\\README.md
[main] WARN io.swagger.codegen.DefaultCodegen - ApiResponse (reserved word) cannot be used as model name. Renamed to ModelApiResponse
[main] WARN io.swagger.codegen.DefaultCodegen - ApiResponse (reserved word) cannot be used as model name. Renamed to ModelApiResponse
[main] WARN io.swagger.codegen.DefaultCodegen - ApiResponse (reserved word) cannot be used as model name. Renamed to ModelApiResponse
[main] WARN io.swagger.codegen.DefaultCodegen - ApiResponse (reserved word) cannot be used as model name. Renamed to ModelApiResponse
[main] WARN io.swagger.codegen.DefaultCodegen - ApiResponse (reserved word) cannot be used as model name. Renamed to ModelApiResponse
[main] WARN io.swagger.codegen.DefaultCodegen - ApiResponse (reserved word) cannot be used as model name. Renamed to ModelApiResponse
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\src\gen\java\io\swagger\model\Category.java
[main] WARN io.swagger.codegen.DefaultCodegen - ApiResponse (reserved word) cannot be used as model name. Renamed to ModelApiResponse
[main] WARN io.swagger.codegen.DefaultCodegen - ApiResponse (reserved word) cannot be used as model name. Renamed to ModelApiResponse
[main] WARN io.swagger.codegen.DefaultCodegen - ApiResponse (reserved word) cannot be used as model name. Renamed to ModelApiResponse
[main] WARN io.swagger.codegen.DefaultCodegen - ApiResponse (reserved word) cannot be used as model name. Renamed to ModelApiResponse
[main] WARN io.swagger.codegen.DefaultCodegen - ApiResponse (reserved word) cannot be used as model name. Renamed to ModelApiResponse
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\src\gen\java\io\swagger\model\ModelApiResponse.java
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\src\gen\java\io\swagger\model\Order.java
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\src\gen\java\io\swagger\model\Pet.java
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\src\gen\java\io\swagger\model\Tag.java
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\src\gen\java\io\swagger\model\User.java
[main] WARN io.swagger.codegen.DefaultCodegen - ApiResponse (reserved word) cannot be used as model name. Renamed to ModelApiResponse
[main] WARN io.swagger.codegen.DefaultCodegen - ApiResponse (reserved word) cannot be used as model name. Renamed to ModelApiResponse
[main] WARN io.swagger.codegen.DefaultCodegen - ApiResponse (reserved word) cannot be used as model name. Renamed to ModelApiResponse
[main] WARN io.swagger.codegen.DefaultCodegen - ApiResponse (reserved word) cannot be used as model name. Renamed to ModelApiResponse
[main] WARN io.swagger.codegen.DefaultCodegen - ApiResponse (reserved word) cannot be used as model name. Renamed to ModelApiResponse
[main] WARN io.swagger.codegen.DefaultCodegen - ApiResponse (reserved word) cannot be used as model name. Renamed to ModelApiResponse
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\src\main\java\io\swagger\api\impl\PetApiServiceImpl.java
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\src\gen\java\io\swagger\api\PetApi.java
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\src\main\java\io\swagger\api\impl\StoreApiServiceImpl.java
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\src\gen\java\io\swagger\api\StoreApi.java
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\src\main\java\io\swagger\api\impl\UserApiServiceImpl.java
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\src\gen\java\io\swagger\api\UserApi.java
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\pom.xml
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\build.gradle
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\settings.gradle
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\src/main/webapp/WEB-INF\web.xml
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\src\main\java/io/swagger/api\JacksonConfig.java
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\src/main/webapp/WEB-INF\jboss-web.xml
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\src\main\java/io/swagger/api\RestApplication.java
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\.swagger-codegen-ignore
[main] INFO io.swagger.codegen.AbstractGenerator - writing file C:\dev\swagger-codegen-fork\samples\server\petstore\jaxrs-resteasy\eap-java8\.swagger-codegen\VERSION
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=256M; support was removed in 8.0
```

As you can see, I have zero trouble generating the code. I guess there is some settings on your local computer that use java 1.5, but I'm no expert in configuration so I can tell where to change it.

@wing328 Can we consider this ready to merge?

daveish commented:
The failing part is not the generator itself, but the Maven verify goal after generating the sample:

`mvn -q clean verify -Psamples`

JFCote commented:
@daveish I have never run such test and never was required to. 
When I try to run that test, I have multiple errors related to the plugin itself (nothing related to java8 or diamonds)

```
$ mvn -q clean verify -Psamples -e
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:1.4.1:enforce (enforce-versions) on project swagger-codegen-project: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed. -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:1.4.1:enforce (enforce-versions) on project swagger-codegen-project: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed.
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:217)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:59)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.singleThreadedBuild(LifecycleStarter.java:183)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:161)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:320)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:156)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:537)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:196)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:141)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:290)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:230)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:409)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:352)
Caused by: org.apache.maven.plugin.MojoExecutionException: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed.
        at org.apache.maven.plugins.enforcer.EnforceMojo.execute(EnforceMojo.java:237)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:101)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:209)
        ... 19 more
[ERROR]
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

I guess I don't have this plugin on my computer or some other configuration that is on the CI.

But there is something I'm sure is that the generated java code (with the <>) is perfectly legal in java8 and should compile.

@wing328 I don't know what is this "verify" command but it doesn't seems to work on my computer and it doesn't seems to work for different reasons on @daveish computers. Do we really need to run this locally?

daveish commented:
The verify goal is run on CircleCI. You made a change for all Java generators, so all samples should have been generated to verify the correctness via CircleCI. I made a change which affected the majority of Java generators, too. Therefore I ran all the sample generators. What happened is, that your code broke the Tests for CircleCI somehow. I know, that Java accepts the diamond operator. 

Did you generate all Java samples when you pushed your changes for this PR?

Regards
David

JFCote commented:
@daveish No, but @wing328 did it. Check above message in the thread:

> @JFCote I've run ./bin/java-petstore-all.sh to update all Java API clients. CircleCI tests look good: https://circleci.com/gh/swagger-api/swagger-codegen/1658

daveish commented:
@JFCote Yes, you are right. I saw that earlier. Unfortunately `./bin/jaxrs-resteasy-eap-java8-petstore-server.sh` is not being executed when running `./bin/java-petstore-all.sh`. Try executing on your machine. You will see that there is a change. And this change produces the error on CircleCI.

wing328 commented on modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/AbstractJavaCodegen.java:
```diff
@@ -639,7 +639,16 @@ public String toDefaultValue(Property p) {
+            if (java8obj != null) {
+                Boolean java8 = Boolean.valueOf(java8obj.toString());
+                if (java8 != null && java8) {
+                    typeDeclaration = "";
```
@JFCote What about returning earlier here with `return String.format(pattern, "")`?

For the return below (line 651), we keep the original one `return String.format(pattern, getTypeDeclaration(ap.getItems()));` Would that work for both Play and RestEasy generator?

cc @daveish

JFCote replied:
@wing328 I can change it if you want, I learn the old school way of not having too much `return` in a function (even if that was only required in the old days). Let me know if you want me to change it.

As for the line 651, it the exact same return as before. It's just that I have declared `typeDeclaration` online 642. We change it to "" only if java8 is set to true.

JFCote commented:
@wing328 Do you want me to do the changes or is this ok? Check my comment in your review. Like I said, this doesn't change anything in the logic, it is only if we want a variable or multiple returns.
As for why it fails for @daveish , I really don't understand.

wing328 commented:
If I add the following to "samples/server/petstore/jaxrs-resteasy/eap-java8/pom.xml" to specify Java version 8, `mvn test` no longer reports any errors:
```
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.6.1</version>
                <configuration>
                    <source>1.8</source>
                    <target>1.8</target>
                </configuration>
            </plugin>
```
 I'll need to perform more tests though to ensure it won't break other Java server generators...

wing328 commented:
I checked other Java server generators (e.g Spring) and found only RESTEasy has this particular issue. I'll merge this PR into master and then apply the fix I developed.

wing328 commented:
@daveish I've pushed the fix via 3b134319fa428a938eba535d7da0120ad5ce5a6b to the latest master

Please pull the latest and see if you still see any issue.
